### ISSUE TITLE
asyncio  ERROR: Could not fetch results due to an unexpected error

### ISSUE THREAD

**Issue by dnbard** [nicbou/burgeramt-appointments-websockets#14]
Good part:
```
[2023-04-15 02:05:06] INFO: Getting appointment URL for https://service.berlin.de/dienstleistung/325881/
[2023-04-15 02:05:08] INFO: URL found: https://service.berlin.de/terminvereinbarung/termin/tag.php?termin=1&anliegen[]=325881&dienstleisterlist=121364,121362&herkunft=http%3A%2F%2Fservice.berlin.de%2Fdienstleistung%2F325881%2F
[2023-04-15 02:05:09] INFO: Server is running on port 80. Looking for appointments every 180 seconds.
```
Exception:
```
[2023-04-15 02:07:25] ERROR: Could not fetch results due to an unexpected error.
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.10/site-packages/appointments/appointments.py", line 110, in look_for_appointments
    appointments = await get_appointments(appointments_url, email, script_id)
  File "/opt/homebrew/lib/python3.10/site-packages/appointments/appointments.py", line 88, in get_appointments
    + parse_appointment_dates(await response_p2.text())
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/client_reqrep.py", line 1081, in text
    await self.read()
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/client_reqrep.py", line 1037, in read
    self._body = await self.content.read()
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/streams.py", line 375, in read
    block = await self.readany()
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/streams.py", line 397, in readany
    await self._wait("readany")
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/streams.py", line 303, in _wait
    with self._timer:
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/helpers.py", line 721, in __exit__
    raise asyncio.TimeoutError from None
asyncio.exceptions.TimeoutError
```

**Comment by nicbou** [comment]
Hm interesting. So it fails while waiting for the second page of the appointments calendar.

1. Does it happen every time?
2. Does it stop the script completely, or does it keep running?

**Comment by dnbard** [comment]
I started the script a couple of minutes ago again: 

```
[2023-04-15 10:29:52] INFO: Getting appointment URL for https://service.berlin.de/dienstleistung/120877/
[2023-04-15 10:29:53] INFO: URL found: https://service.berlin.de/terminvereinbarung/termin/tag.php?termin=1&anliegen[]=120877&dienstleisterlist=122210,122217,122219,122227,122231,122238,122243,122254,122252,122260,122262,122271,122273,122277,122280,122282,122284,122291,122285,122286,122296,150230,122297,122294,122312,122314,122304,122311,122309,122281,122279,122283,122276,122274,122267,122246,122251,122257,122208,122226&herkunft=http%3A%2F%2Fservice.berlin.de%2Fdienstleistung%2F120877%2F
[2023-04-15 10:29:53] INFO: Server is running on port 80. Looking for appointments every 180 seconds.
[2023-04-15 10:29:58] INFO: Found 0 appointments: []
[2023-04-15 10:33:24] ERROR: Could not fetch results due to an unexpected error.
// same traceback
```

1. Yes. I have tested it with 3 different "services you want an appointment for". 
2. It keeps running:
```
[2023-04-15 10:36:29] INFO: Found 0 appointments: []
```
After that point, no logs were produced, but sound still playing once script find the appointment. I waited from 10.29 to 10.54, heard sound couple of times and was able to get the appointment.

**Comment by dnbard** [comment]
Update: After I interrupted the script (Control + C), it printed all missing logs to terminal:
```
[2023-04-15 10:29:52] INFO: Getting appointment URL for https://service.berlin.de/dienstleistung/120877/
[2023-04-15 10:29:53] INFO: URL found: https://service.berlin.de/terminvereinbarung/termin/tag.php?termin=1&anliegen[]=120877&dienstleisterlist=122210,122217,122219,122227,122231,122238,122243,122254,122252,122260,122262,122271,122273,122277,122280,122282,122284,122291,122285,122286,122296,150230,122297,122294,122312,122314,122304,122311,122309,122281,122279,122283,122276,122274,122267,122246,122251,122257,122208,122226&herkunft=http%3A%2F%2Fservice.berlin.de%2Fdienstleistung%2F120877%2F
[2023-04-15 10:29:53] INFO: Server is running on port 80. Looking for appointments every 180 seconds.
[2023-04-15 10:29:58] INFO: Found 0 appointments: []
[2023-04-15 10:33:24] ERROR: Could not fetch results due to an unexpected error.
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.10/site-packages/appointments/appointments.py", line 110, in look_for_appointments
    appointments = await get_appointments(appointments_url, email, script_id)
  File "/opt/homebrew/lib/python3.10/site-packages/appointments/appointments.py", line 88, in get_appointments
    + parse_appointment_dates(await response_p2.text())
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/client_reqrep.py", line 1081, in text
    await self.read()
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/client_reqrep.py", line 1037, in read
    self._body = await self.content.read()
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/streams.py", line 375, in read
    block = await self.readany()
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/streams.py", line 397, in readany
    await self._wait("readany")
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/streams.py", line 303, in _wait
    with self._timer:
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/helpers.py", line 721, in __exit__
    raise asyncio.TimeoutError from None
asyncio.exceptions.TimeoutError
[2023-04-15 10:36:29] INFO: Found 0 appointments: []
[2023-04-15 10:39:34] INFO: Found 0 appointments: []
[2023-04-15 10:42:39] INFO: Found 1 appointments: ['2023-04-19T00:00:00Z']
[2023-04-15 10:46:03] ERROR: Could not fetch results due to an unexpected error.
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.10/site-packages/appointments/appointments.py", line 110, in look_for_appointments
    appointments = await get_appointments(appointments_url, email, script_id)
  File "/opt/homebrew/lib/python3.10/site-packages/appointments/appointments.py", line 88, in get_appointments
    + parse_appointment_dates(await response_p2.text())
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/client_reqrep.py", line 1081, in text
    await self.read()
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/client_reqrep.py", line 1037, in read
    self._body = await self.content.read()
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/streams.py", line 375, in read
    block = await self.readany()
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/streams.py", line 397, in readany
    await self._wait("readany")
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/streams.py", line 303, in _wait
    with self._timer:
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/helpers.py", line 721, in __exit__
    raise asyncio.TimeoutError from None
asyncio.exceptions.TimeoutError
[2023-04-15 10:49:26] ERROR: Could not fetch results due to an unexpected error.
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.10/site-packages/appointments/appointments.py", line 110, in look_for_appointments
    appointments = await get_appointments(appointments_url, email, script_id)
  File "/opt/homebrew/lib/python3.10/site-packages/appointments/appointments.py", line 88, in get_appointments
    + parse_appointment_dates(await response_p2.text())
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/client_reqrep.py", line 1081, in text
    await self.read()
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/client_reqrep.py", line 1037, in read
    self._body = await self.content.read()
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/streams.py", line 375, in read
    block = await self.readany()
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/streams.py", line 397, in readany
    await self._wait("readany")
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/streams.py", line 303, in _wait
    with self._timer:
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/helpers.py", line 721, in __exit__
    raise asyncio.TimeoutError from None
asyncio.exceptions.TimeoutError
[2023-04-15 10:52:49] ERROR: Could not fetch results due to an unexpected error.
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.10/site-packages/appointments/appointments.py", line 110, in look_for_appointments
    appointments = await get_appointments(appointments_url, email, script_id)
  File "/opt/homebrew/lib/python3.10/site-packages/appointments/appointments.py", line 88, in get_appointments
    + parse_appointment_dates(await response_p2.text())
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/client_reqrep.py", line 1081, in text
    await self.read()
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/client_reqrep.py", line 1037, in read
    self._body = await self.content.read()
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/streams.py", line 375, in read
    block = await self.readany()
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/streams.py", line 397, in readany
    await self._wait("readany")
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/streams.py", line 303, in _wait
    with self._timer:
  File "/opt/homebrew/lib/python3.10/site-packages/aiohttp/helpers.py", line 721, in __exit__
    raise asyncio.TimeoutError from None
asyncio.exceptions.TimeoutError
[2023-04-15 10:55:54] INFO: Found 1 appointments: ['2023-04-19T00:00:00Z']
^CTraceback (most recent call last):
  File "/opt/homebrew/bin/appointments", line 67, in <module>
    asyncio.run(watch_for_appointments(service_page_url, email, args.id, args.port, args.quiet))
  File "/opt/homebrew/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/opt/homebrew/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 636, in run_until_complete
    self.run_forever()
  File "/opt/homebrew/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 603, in run_forever
    self._run_once()
  File "/opt/homebrew/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 1871, in _run_once
    event_list = self._selector.select(timeout)
  File "/opt/homebrew/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/selectors.py", line 562, in select
    kev_list = self._selector.control(None, max_ev, timeout)
KeyboardInterrupt
```

**Comment by dnbard** [comment]
And I have tried once again. It not always happens on second iteration:

```
[2023-04-15 11:05:27] INFO: Server is running on port 80. Looking for appointments every 180 seconds.
[2023-04-15 11:05:30] INFO: Found 1 appointments: ['2023-04-21T00:00:00Z']
[2023-04-15 11:08:33] INFO: Found 0 appointments: []
[2023-04-15 11:11:37] INFO: Found 0 appointments: []
[2023-04-15 11:14:40] INFO: Found 0 appointments: []
[2023-04-15 11:18:05] ERROR: Could not fetch results due to an unexpected error.
```

**Comment by nicbou** [comment]
Okay, so this is the same issue I have encountered last week. It's due to the city tightening their firewall rules [after a recent cyberattack](https://twitter.com/aboutberlin/status/1643632257242918917?cxt=HHwWisCzpYy1rc8tAAAA). The symptoms are timeouts and dropped connections. My [Wachete](https://www.wachete.com/) watchers also fail since then.

We can't change the script to fix this, unfortunately. The script on All About Berlin works fine since I reported the issue, so it might be a manual exemption.

I'll leave this ticket open as the script should handle these errors more elegantly, with human-readable error messages. I will not work on this right now, but feel free to submit a pull request.
